crates/services/executor/src/executor.rs

```diff
@@ -1030,228 +1080,671 @@ where
-                *mint.mint_amount(),
-            )
-            .map_err(|e| format!("{e}"))
-            .map_err(ExecutorError::CoinbaseCannotIncreaseBalance)?;
-            sub_block_db_commit.commit()?;
+    fn convert_maybe_checked_tx_to_checked_tx(
+        &self,
+        tx: MaybeCheckedTransaction,
+        header: &PartialBlockHeader,
+    ) -> ExecutorResult<CheckedTransaction> {
+        let block_height = *header.height();
+        let checked_tx = match tx {
+            MaybeCheckedTransaction::Transaction(tx) => tx
+                .into_checked_basic(block_height, &self.consensus_params)?
+                .into(),
+            MaybeCheckedTransaction::CheckedTransaction(checked_tx) => checked_tx,
+        };
+        Ok(checked_tx)
+    }
+
+    #[allow(clippy::too_many_arguments)]
+    fn execute_mint<T>(
+        &self,
+        checked_mint: Checked<Mint>,
+        header: &PartialBlockHeader,
+        coinbase_contract_id: ContractId,
+        gas_price: Word,
+        execution_data: &mut ExecutionData,
+        block_st_transaction: &mut StorageTransaction<T>,
+    ) -> ExecutorResult<Transaction>
+    where
+        T: KeyValueInspect<Column = Column>,
+    {
+        execution_data.found_mint = true;
+
+        Self::check_mint_has_expected_index(&checked_mint, execution_data)?;
+
+        let coinbase_id = checked_mint.id();
+        let (mut mint, _) = checked_mint.into();
+
+        if mint.input_contract().contract_id == ContractId::zeroed() {
+            Self::verify_mint_for_empty_contract(&mint)?;
+        } else {
+            Self::check_mint_amount(&mint, execution_data.coinbase)?;
+            Self::check_gas_price(&mint, gas_price)?;
+
+            let input = mint.input_contract().clone();
+            let mut inputs = [Input::Contract(input)];
+
+            if self.options.utxo_validation {
+                self.verify_inputs_exist(
+                    block_st_transaction,
+                    inputs.as_mut_slice(),
+                    header.da_height,
+                )?;
+            }
+
+            self.compute_inputs(inputs.as_mut_slice(), block_st_transaction)?;
 
-            self.persist_output_utxos(
-                block_height,
+            let (input, output) = self.execute_mint_with_vm(
+                header,
+                coinbase_contract_id,
                 execution_data,
-                &coinbase_id,
                 block_st_transaction,
-                inputs.as_slice(),
-                outputs.as_slice(),
+                &coinbase_id,
+                &mut mint,
+                &mut inputs,
             )?;
-            self.compute_state_of_not_utxo_outputs(
-                outputs.as_mut_slice(),
-                inputs.as_slice(),
+
+            *mint.input_contract_mut() = input;
+            *mint.output_contract_mut() = output;
+        }
+
+        Self::store_mint_tx(mint, execution_data, coinbase_id, block_st_transaction)
+    }
+
+    fn validate_mint<T>(
```

Comment:
As I said in the past, validation of the block is equal to producing this block from the validator's perspective and comparing it.

It is the main idea of what we need to follow. The `execute_mint` and `validate_mint` should be the same and do the same steps.

You can compare the produced mint transaction at the end of the `validate_block` function to be sure that it was created correctly.